transports/quic/src/stream_map.rs

```diff
@@ -0,0 +1,522 @@
+    }
+
+    fn finish(&mut self, id: &StreamId) -> Result<(), quinn_proto::FinishError> {
+        self.connection.finish(id.0).map_err(|e| {
+            self.get(&id).finish();
+            e
+        })
+    }
+
+    pub(crate) fn write(
+        &mut self,
+        cx: &mut Context<'_>,
+        substream: &StreamId,
+        buf: &[u8],
+    ) -> Result<usize, quinn_proto::WriteError> {
+        use quinn_proto::WriteError;
+        self.wake_driver();
+        self.connection.write(substream.0, buf).map_err(|e| {
+            let stream = self.get(substream);
+            match e {
+                WriteError::Blocked => stream.set_writer(cx.waker().clone()),
+                WriteError::Stopped(_) | WriteError::UnknownStream => stream.finish(),
+            }
+            e
+        })
+    }
+
+    pub(crate) fn side(&self) -> quinn_proto::Side {
+        self.connection.side()
+    }
+
+    pub(crate) fn read(
+        &mut self,
+        cx: &mut Context<'_>,
+        substream: &StreamId,
+        buf: &mut [u8],
+    ) -> Result<usize, quinn_proto::ReadError> {
+        use quinn_proto::ReadError;
+        self.wake_driver();
+        match self.connection.read(substream.0, buf) {
+            e @ Err(ReadError::Blocked) => {
+                self.get(substream).set_reader(cx.waker().clone());
+                e
+            }
+            e => e,
+        }
+    }
+
+    pub(crate) fn get_pending_stream(&mut self) -> Either<StreamId, oneshot::Receiver<StreamId>> {
+        self.wake_driver();
+        let pending = std::mem::replace(&mut self.pending_stream, None);
+        match pending.or_else(|| self.open_stream()) {
+            Some(stream) => Either::Left(stream),
+            None => {
+                let (sender, receiver) = oneshot::channel();
+                self.connectors.push_front(sender);
+                self.wake_driver();
+                Either::Right(receiver)
+            }
+        }
+    }
+
+    pub(crate) fn close_reason(&self) -> Result<(), Error> {
+        match self
+            .close_reason
+            .as_ref()
+            .map(|e| Error::ConnectionError(e.clone()))
+        {
+            Some(e) => Err(e),
+            None => Ok(()),
+        }
+    }
+
+    pub(crate) fn destroy_stream(&mut self, stream: StreamId) {
+        trace!("Removing substream {:?} from map", *stream);
+        self.connection.destroy_stream(*stream);
+        self.map.remove(&stream.0).expect(
+            "Internal state corrupted. \
+             You probably used a Substream with the wrong StreamMuxer",
+        );
+    }
+
+    pub(crate) fn close(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        trace!(
+            "close() called with {} outbound streams for side {:?}",
+            self.connection.send_streams(),
+            self.connection.side()
+        );
+        self.wake_closer();
+        self.close_waker = Some(cx.waker().clone());
```

Comment:
What is the purpose of `close_waker`? None of the methods below return `Poll` so why not just call them? Should polling not simply continue until the expected number of `StreamFinished` events have been received?

Reply:
`close_waker` is awoken when a `StreamFinished` event is received or the stream is lost, so that this task can be woken up.  There is no other way of doing so that I can think of.